Improve `invalid_reference_casting` lint

Urgau opened the pull request:
This PR is a follow-up to https://github.com/rust-lang/rust/pull/111567 and https://github.com/rust-lang/rust/pull/113422.

This PR does multiple things:
 - First it adds support for deferred de-reference, the goal is to support code like this, where the casting and de-reference are not done on the same expression
    ```rust
    let myself = self as *const Self as *mut Self;
    *myself = Self::Ready(value);
    ```
 - Second it does not lint anymore on SB/TB UB code by only checking assignments (`=`, `+=`, ...) and creation of mutable references `&mut *`
 - Thirdly it greatly improves the diagnostics in particular for cast from `&mut` to `&mut` or assignments
 - ~~And lastly it renames the lint from `cast_ref_to_mut` to `invalid_reference_casting` which is more consistent with the ["rules"](https://github.com/rust-lang/rust-clippy/issues/2845) and also more consistent with what the lint checks~~ *https://github.com/rust-lang/rust/pull/113422*

This PR is best reviewed commit by commit.

r? compiler

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5214881123/jobs/9411715446) [(plain)](https://github.com/rust-lang/rust/commit/5291f326972b26428199a434aaead0a500021f64/checks/14116591809/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] tests/ui/const-generics/issues/issue-100313.rs stdout ----
diff of stderr:

- error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
+ error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
3    |
3    |
4 LL |             *(B as *const bool as *mut bool) = false;
-    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
6    |
-    = note: `#[deny(cast_ref_to_mut)]` on by default
-    = note: `#[deny(cast_ref_to_mut)]` on by default
+    = note: `#[deny(invalid_reference_casting)]` on by default
9 error[E0080]: evaluation of constant value failed
10   --> $DIR/issue-100313.rs:10:13



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/issues/issue-100313/issue-100313.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args const-generics/issues/issue-100313.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/const-generics/issues/issue-100313.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/issues/issue-100313" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/issues/issue-100313/auxiliary"
stdout: none
--- stderr -------------------------------
error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
  --> fake-test-src-base/const-generics/issues/issue-100313.rs:10:13
   |
LL |             *(B as *const bool as *mut bool) = false;
   |
   |
   = note: `#[deny(invalid_reference_casting)]` on by default
error[E0080]: evaluation of constant value failed
  --> fake-test-src-base/const-generics/issues/issue-100313.rs:10:13
   |
   |
LL |             *(B as *const bool as *mut bool) = false;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ writing to alloc7 which is read-only
   |
note: inside `T::<&true>::set_false`
  --> fake-test-src-base/const-generics/issues/issue-100313.rs:10:13
   |
LL |             *(B as *const bool as *mut bool) = false;
note: inside `_`
  --> fake-test-src-base/const-generics/issues/issue-100313.rs:19:5
   |
   |
LL |     x.set_false();

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0080`.
```

</details>

Urgau committed:
Adjust some tests for invalid_reference_casting improvements

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5215068415/jobs/9412136127) [(plain)](https://github.com/rust-lang/rust/commit/8a01d5f9e3703171dfbecd783f0698645c48a068/checks/14117149686/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
...........................

failures:

---- src/cell.rs - cell::UnsafeCell (line 1896) stdout ----
error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
  |
6 |   let t = ptr as *const UnsafeCell<T> as *mut T;
  |           ------------------------------------- casting happend here
...
...
9 |   unsafe { &mut *t }
  |
  |
  = note: `#[deny(invalid_reference_casting)]` on by default
error: aborting due to previous error

Couldn't compile the test.


failures:
    src/cell.rs - cell::UnsafeCell (line 1896)

test result: FAILED. 4121 passed; 1 failed; 41 ignored; 0 measured; 0 filtered out; finished in 61.91s

error: doctest failed, to rerun pass `-p core --doc`
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #113151) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5414133378/jobs/9840737053) [(plain)](https://github.com/rust-lang/rust/commit/1f84677cc3e0cda6eb7e0787d665e8a2649db307/checks/14661378294/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

## Running ui tests in tests/fail against miri for x86_64-unknown-linux-gnu
   Compiler: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri "--error-format=json" "-Dwarnings" "-Dunused" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" OUT_DIR
   Building test dependencies...
.............................................................FF................................FF... 100
.................................................................................................... 300
.................................................................................................... 400
................................
tests/fail/both_borrows/illegal_write1.rs (revision `stack`) FAILED:
tests/fail/both_borrows/illegal_write1.rs (revision `stack`) FAILED:
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--extern" "getrandom=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-8197e2dce1dfb558.rlib" "--extern" "getrandom=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-8197e2dce1dfb558.rmeta" "--extern" "getrandom_1=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-19aeac02a6c0ffc3.rlib" "--extern" "getrandom_1=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-19aeac02a6c0ffc3.rmeta" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-ef2fbe160fcdd653.rlib" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-ef2fbe160fcdd653.rmeta" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-54bb73fbdf6df9ac.rlib" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-54bb73fbdf6df9ac.rmeta" "--extern" "rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/librand-0ff886e7eaab9922.rlib" "--extern" "rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/librand-0ff886e7eaab9922.rmeta" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-6a0f79efa3d70627.rlib" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-6a0f79efa3d70627.rmeta" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-189fb0628764d2b2.rlib" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-189fb0628764d2b2.rmeta" "--extern" "miri_test_deps=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/miri-test-deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/memchr-8c21ca7640ca0068" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/getrandom-3c088a94c925444e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/quote-059f15a9694b54ae" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/libc-117bf3c154c8368e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/proc-macro2-66abfe3863feaa0b" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/log-1ec6f959fa319c8e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/lock_api-67529c054bcce400" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/tokio-644c63a1da1ce210" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/parking_lot_core-057742f3b212972a" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/syn-2407d38215cd42b2" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui" "tests/fail/both_borrows/illegal_write1.rs" "--cfg=stack" "--edition" "2021"





actual output differed from expected
Execute `./miri bless` to update `tests/fail/both_borrows/illegal_write1.stack.stderr` to the actual output
--- tests/fail/both_borrows/illegal_write1.stack.stderr
+++ <stderr output>
-error: Undefined Behavior: attempting a write access using <TAG> at ALLOC[0x0], but that tag only grants SharedReadOnly permission for this location
+error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
   --> $DIR/illegal_write1.rs:LL:CC
    |
+LL |         let x: *mut u32 = xref as *const _ as *mut _;
+   |                           -------------------------- casting happend here
 LL |         unsafe { *x = 42 };
-   |                  |
-   |                  |
-   |                  attempting a write access using <TAG> at ALLOC[0x0], but that tag only grants SharedReadOnly permission for this location
-   |                  this error occurs as part of an access at ALLOC[0x0..0x4]
-   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
-   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
-   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
-help: <TAG> was created by a SharedReadOnly retag at offsets [0x0..0x4]
-  --> $DIR/illegal_write1.rs:LL:CC
-   |
-LL |         let x: *mut u32 = xref as *const _ as *mut _;
-   = note: BACKTRACE (of the first span):
-   = note: inside `main` at $DIR/illegal_write1.rs:LL:CC
-   = note: inside `main` at $DIR/illegal_write1.rs:LL:CC
+   = note: `#[deny(invalid_reference_casting)]` on by default
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+error: miri cannot be run on programs that fail compilation
~error: aborting due to 2 previous errors
 
 



`/write access .* tag only grants SharedReadOnly permission/` does not match stderr output

There were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
There were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: miri cannot be run on programs that fail compilation
There were 1 unmatched diagnostics at tests/fail/both_borrows/illegal_write1.rs:9
There were 1 unmatched diagnostics at tests/fail/both_borrows/illegal_write1.rs:9
    Error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
full stderr:
full stderr:
error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
##[error]  --> tests/fail/both_borrows/illegal_write1.rs:9:18
   |
LL |         let x: *mut u32 = xref as *const _ as *mut _;
   |                           -------------------------- casting happend here
LL |         unsafe { *x = 42 };
   |
   |
   = note: `#[deny(invalid_reference_casting)]` on by default

error: miri cannot be run on programs that fail compilation
error: aborting due to 2 previous errors





tests/fail/both_borrows/illegal_write1.rs (revision `tree`) FAILED:
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--extern" "getrandom=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-8197e2dce1dfb558.rlib" "--extern" "getrandom=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-8197e2dce1dfb558.rmeta" "--extern" "getrandom_1=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-19aeac02a6c0ffc3.rlib" "--extern" "getrandom_1=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-19aeac02a6c0ffc3.rmeta" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-ef2fbe160fcdd653.rlib" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-ef2fbe160fcdd653.rmeta" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-54bb73fbdf6df9ac.rlib" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-54bb73fbdf6df9ac.rmeta" "--extern" "rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/librand-0ff886e7eaab9922.rlib" "--extern" "rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/librand-0ff886e7eaab9922.rmeta" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-6a0f79efa3d70627.rlib" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-6a0f79efa3d70627.rmeta" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-189fb0628764d2b2.rlib" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-189fb0628764d2b2.rmeta" "--extern" "miri_test_deps=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/miri-test-deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/memchr-8c21ca7640ca0068" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/getrandom-3c088a94c925444e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/quote-059f15a9694b54ae" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/libc-117bf3c154c8368e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/proc-macro2-66abfe3863feaa0b" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/log-1ec6f959fa319c8e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/lock_api-67529c054bcce400" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/tokio-644c63a1da1ce210" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/parking_lot_core-057742f3b212972a" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/syn-2407d38215cd42b2" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui" "tests/fail/both_borrows/illegal_write1.rs" "--cfg=tree" "-Zmiri-tree-borrows" "--edition" "2021"





actual output differed from expected
Execute `./miri bless` to update `tests/fail/both_borrows/illegal_write1.tree.stderr` to the actual output
--- tests/fail/both_borrows/illegal_write1.tree.stderr
+++ <stderr output>
-error: Undefined Behavior: write access through <TAG> is forbidden
+error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
   --> $DIR/illegal_write1.rs:LL:CC
    |
+LL |         let x: *mut u32 = xref as *const _ as *mut _;
+   |                           -------------------------- casting happend here
 LL |         unsafe { *x = 42 };
~   |                  ^^^^^^^ write access through <TAG> is forbidden
-   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Tree Borrows rules it violated are still experimental
-   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Tree Borrows rules it violated are still experimental
-   = help: the accessed tag <TAG> has state Frozen which forbids this child write access
-help: the accessed tag <TAG> was created here, in the initial state Frozen
-  --> $DIR/illegal_write1.rs:LL:CC
-   |
-LL |     let xref = &*target;
-   = note: BACKTRACE (of the first span):
-   = note: inside `main` at $DIR/illegal_write1.rs:LL:CC
-   = note: inside `main` at $DIR/illegal_write1.rs:LL:CC
+   = note: `#[deny(invalid_reference_casting)]` on by default
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+error: miri cannot be run on programs that fail compilation
~error: aborting due to 2 previous errors
 
 



`/write access through .* is forbidden/` does not match stderr output

There were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
There were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: miri cannot be run on programs that fail compilation
There were 1 unmatched diagnostics at tests/fail/both_borrows/illegal_write1.rs:9
There were 1 unmatched diagnostics at tests/fail/both_borrows/illegal_write1.rs:9
    Error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
full stderr:
full stderr:
error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
##[error]  --> tests/fail/both_borrows/illegal_write1.rs:9:18
   |
LL |         let x: *mut u32 = xref as *const _ as *mut _;
   |                           -------------------------- casting happend here
LL |         unsafe { *x = 42 };
   |
   |
   = note: `#[deny(invalid_reference_casting)]` on by default

error: miri cannot be run on programs that fail compilation
error: aborting due to 2 previous errors





tests/fail/both_borrows/shr_frozen_violation1.rs (revision `stack`) FAILED:
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--extern" "getrandom=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-8197e2dce1dfb558.rlib" "--extern" "getrandom=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-8197e2dce1dfb558.rmeta" "--extern" "getrandom_1=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-19aeac02a6c0ffc3.rlib" "--extern" "getrandom_1=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-19aeac02a6c0ffc3.rmeta" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-ef2fbe160fcdd653.rlib" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-ef2fbe160fcdd653.rmeta" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-54bb73fbdf6df9ac.rlib" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-54bb73fbdf6df9ac.rmeta" "--extern" "rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/librand-0ff886e7eaab9922.rlib" "--extern" "rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/librand-0ff886e7eaab9922.rmeta" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-6a0f79efa3d70627.rlib" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-6a0f79efa3d70627.rmeta" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-189fb0628764d2b2.rlib" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-189fb0628764d2b2.rmeta" "--extern" "miri_test_deps=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/miri-test-deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/memchr-8c21ca7640ca0068" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/getrandom-3c088a94c925444e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/quote-059f15a9694b54ae" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/libc-117bf3c154c8368e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/proc-macro2-66abfe3863feaa0b" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/log-1ec6f959fa319c8e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/lock_api-67529c054bcce400" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/tokio-644c63a1da1ce210" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/parking_lot_core-057742f3b212972a" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/syn-2407d38215cd42b2" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui" "tests/fail/both_borrows/shr_frozen_violation1.rs" "--cfg=stack" "--edition" "2021"






actual output differed from expected
Execute `./miri bless` to update `tests/fail/both_borrows/shr_frozen_violation1.stack.stderr` to the actual output
--- tests/fail/both_borrows/shr_frozen_violation1.stack.stderr
+++ <stderr output>
-error: Undefined Behavior: attempting a write access using <TAG> at ALLOC[0x0], but that tag only grants SharedReadOnly permission for this location
+error: unknown lint: `cast_ref_to_mut`
    |
    |
-LL |         *(x as *const i32 as *mut i32) = 7;
-   |         |
-   |         |
-   |         attempting a write access using <TAG> at ALLOC[0x0], but that tag only grants SharedReadOnly permission for this location
-   |         this error occurs as part of an access at ALLOC[0x0..0x4]
+LL | #![allow(cast_ref_to_mut)]
    |
-   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
-   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
-   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
-help: <TAG> was created by a SharedReadOnly retag at offsets [0x0..0x4]
+   = note: `-D unknown-lints` implied by `-D warnings`
+
+error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
    |
    |
 LL |         *(x as *const i32 as *mut i32) = 7;
-   = note: BACKTRACE (of the first span):
-   = note: inside `unknown_code` at $DIR/shr_frozen_violation1.rs:LL:CC
-note: inside `foo`
-  --> $DIR/shr_frozen_violation1.rs:LL:CC
-  --> $DIR/shr_frozen_violation1.rs:LL:CC
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
-LL |     unknown_code(&*x);
-note: inside `main`
-  --> $DIR/shr_frozen_violation1.rs:LL:CC
-   |
-   |
-LL |     println!("{}", foo(&mut 0));
-   |                    ^^^^^^^^^^^
+   = note: `#[deny(invalid_reference_casting)]` on by default
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+error: miri cannot be run on programs that fail compilation
~error: aborting due to 3 previous errors
 
 



`/write access .* only grants SharedReadOnly permission/` does not match stderr output

There were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
There were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: miri cannot be run on programs that fail compilation
There were 1 unmatched diagnostics at tests/fail/both_borrows/shr_frozen_violation1.rs:4
    Error: unknown lint: `cast_ref_to_mut`

There were 1 unmatched diagnostics at tests/fail/both_borrows/shr_frozen_violation1.rs:18
There were 1 unmatched diagnostics at tests/fail/both_borrows/shr_frozen_violation1.rs:18
    Error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
full stderr:
error: unknown lint: `cast_ref_to_mut`
##[error]  --> tests/fail/both_borrows/shr_frozen_violation1.rs:4:10
   |
   |
LL | #![allow(cast_ref_to_mut)]
   |          ^^^^^^^^^^^^^^^
   |
   = note: `-D unknown-lints` implied by `-D warnings`

error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
##[error]  --> tests/fail/both_borrows/shr_frozen_violation1.rs:18:9
   |
LL |         *(x as *const i32 as *mut i32) = 7;
   |
   |
   = note: `#[deny(invalid_reference_casting)]` on by default

error: miri cannot be run on programs that fail compilation
error: aborting due to 3 previous errors





tests/fail/both_borrows/shr_frozen_violation1.rs (revision `tree`) FAILED:
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--extern" "getrandom=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-8197e2dce1dfb558.rlib" "--extern" "getrandom=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-8197e2dce1dfb558.rmeta" "--extern" "getrandom_1=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-19aeac02a6c0ffc3.rlib" "--extern" "getrandom_1=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-19aeac02a6c0ffc3.rmeta" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-ef2fbe160fcdd653.rlib" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-ef2fbe160fcdd653.rmeta" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-54bb73fbdf6df9ac.rlib" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-54bb73fbdf6df9ac.rmeta" "--extern" "rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/librand-0ff886e7eaab9922.rlib" "--extern" "rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/librand-0ff886e7eaab9922.rmeta" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-6a0f79efa3d70627.rlib" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-6a0f79efa3d70627.rmeta" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-189fb0628764d2b2.rlib" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-189fb0628764d2b2.rmeta" "--extern" "miri_test_deps=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/miri-test-deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/memchr-8c21ca7640ca0068" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/getrandom-3c088a94c925444e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/quote-059f15a9694b54ae" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/libc-117bf3c154c8368e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/proc-macro2-66abfe3863feaa0b" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/log-1ec6f959fa319c8e" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/lock_api-67529c054bcce400" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/tokio-644c63a1da1ce210" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/x86_64-unknown-linux-gnu/debug" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/parking_lot_core-057742f3b212972a" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui/miri/debug/build/syn-2407d38215cd42b2" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui" "tests/fail/both_borrows/shr_frozen_violation1.rs" "--cfg=tree" "-Zmiri-tree-borrows" "--edition" "2021"






actual output differed from expected
Execute `./miri bless` to update `tests/fail/both_borrows/shr_frozen_violation1.tree.stderr` to the actual output
--- tests/fail/both_borrows/shr_frozen_violation1.tree.stderr
+++ <stderr output>
-error: Undefined Behavior: write access through <TAG> is forbidden
+error: unknown lint: `cast_ref_to_mut`
    |
    |
-LL |         *(x as *const i32 as *mut i32) = 7;
-   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ write access through <TAG> is forbidden
+LL | #![allow(cast_ref_to_mut)]
    |
-   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Tree Borrows rules it violated are still experimental
-   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Tree Borrows rules it violated are still experimental
-   = help: the accessed tag <TAG> is a child of the conflicting tag <TAG>
-   = help: the conflicting tag <TAG> has state Frozen which forbids this child write access
-help: the accessed tag <TAG> was created here
+   = note: `-D unknown-lints` implied by `-D warnings`
+
+error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
    |
-LL | fn unknown_code(x: &i32) {
-   |                 ^
-   |                 ^
-help: the conflicting tag <TAG> was created here, in the initial state Frozen
-  --> $DIR/shr_frozen_violation1.rs:LL:CC
+LL |         *(x as *const i32 as *mut i32) = 7;
    |
    |
-LL |     unknown_code(&*x);
-   = note: BACKTRACE (of the first span):
-   = note: inside `unknown_code` at $DIR/shr_frozen_violation1.rs:LL:CC
-note: inside `foo`
-  --> $DIR/shr_frozen_violation1.rs:LL:CC
-  --> $DIR/shr_frozen_violation1.rs:LL:CC
-   |
-LL |     unknown_code(&*x);
-note: inside `main`
-  --> $DIR/shr_frozen_violation1.rs:LL:CC
-   |
-   |
-LL |     println!("{}", foo(&mut 0));
-   |                    ^^^^^^^^^^^
+   = note: `#[deny(invalid_reference_casting)]` on by default
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
-note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
+error: miri cannot be run on programs that fail compilation
~error: aborting due to 3 previous errors
 
 



`/write access through .* is forbidden/` does not match stderr output

There were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
There were 1 unmatched diagnostics that occurred outside the testfile and had no pattern
    Error: miri cannot be run on programs that fail compilation
There were 1 unmatched diagnostics at tests/fail/both_borrows/shr_frozen_violation1.rs:4
    Error: unknown lint: `cast_ref_to_mut`

There were 1 unmatched diagnostics at tests/fail/both_borrows/shr_frozen_violation1.rs:18
There were 1 unmatched diagnostics at tests/fail/both_borrows/shr_frozen_violation1.rs:18
    Error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
full stderr:
error: unknown lint: `cast_ref_to_mut`
##[error]  --> tests/fail/both_borrows/shr_frozen_violation1.rs:4:10
   |
   |
LL | #![allow(cast_ref_to_mut)]
   |          ^^^^^^^^^^^^^^^
   |
   = note: `-D unknown-lints` implied by `-D warnings`

error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
##[error]  --> tests/fail/both_borrows/shr_frozen_violation1.rs:18:9
   |
LL |         *(x as *const i32 as *mut i32) = 7;
   |
   |
   = note: `#[deny(invalid_reference_casting)]` on by default

error: miri cannot be run on programs that fail compilation
error: aborting due to 3 previous errors



---
test result: FAIL. 4 tests failed, 425 tests passed, 3 ignored, 0 filtered out
Error: 
   0: tests failed

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/compiletest-4dba848efea0f172 --quiet` (exit status: 1)
Build completed unsuccessfully in 0:03:38
```

</details>

eholk commented on tests/ui/lint/cast_ref_to_mut.rs:
```diff
@@ -10,41 +10,77 @@ extern "C" {
+    *(num as *const i32 as *mut i32);
+    println!("{}", *(num as *const _ as *const i16));
+    println!("{}", *(mut_num as *mut _ as *mut i16));
+    // ffi(a.as_ptr() as *mut _);
```
Why is this commented out? Is it a case that's not handled correctly yet? If so, could you add a comment explaining what's wrong?

Urgau replied:
This was taken from the uplifted clippy test. It's just missing an `a` variable to work.

Urgau replied:
Nevermind, just forgot to re-add `a`. Done now.

eholk commented:
For the SB comments, is the idea that these are UB per Stacked Borrows, but not necessarily _actually_ UB? For example, are these examples allowed under Tree Borrows?

Urgau commented:
> For the SB comments, is the idea that these are UB per Stacked Borrows, but not necessarily _actually_ UB?

Yes, those are UB under Stacked Borrows. As to when ever those are _actually_ UB, I have no idea.

> For example, are these examples allowed under Tree Borrows?

Maybe, I don't know and it shouldn't matter. Neither Stacked Borrows or Tree Borrows are official, so we shouldn't strive for either of them.

----

There is another thing to take into account, when uplifting the lint T-lang only officially approved linting for `&T -> &mut T` casts, so checking for anything more is already borderline.

With this PR only `&T -> &mut T` and `*(&T -> &mut T) = ...` are linted against instead of everything that de-reference the result of an immutable-ref to (mut/const) pointer casting.

WaffleLapkin commented on compiler/rustc_lint/src/reference_casting.rs:
```diff
@@ -0,0 +1,121 @@
+impl<'tcx> LateLintPass<'tcx> for InvalidReferenceCasting {
+    fn check_stmt(&mut self, cx: &LateContext<'tcx>, stmt: &'tcx rustc_hir::Stmt<'tcx>) {
+        let StmtKind::Local(local) = stmt.kind else { return; };
+        let None = local.els else { return; };
```
This should just be if local.els.is_some() { return; }; imo.

bors commented:
:umbrella: The latest upstream changes (presumably #113260) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Urgau commented:
r? @Nilstrieb (per Zulip discussion)

Nilstrieb commented on library/core/src/cell.rs:
````diff
@@ -1893,7 +1893,8 @@ impl<T: ?Sized + fmt::Display> fmt::Display for RefMut<'_, T> {
 ///
-/// ```rust,no_run
+#[cfg_attr(bootstrap, doc = "```rust,no_run")]
+#[cfg_attr(not(bootstrap), doc = "```rust,compile_fail")]
````
This is an interesting case. It's a UB example but it's also something that's fine in both SB and TB as they don't care about `get` at all, so it seems likely to be fine in the future. Avoiding to lint here is quite easy by just checking whether the pointee has interior mutability with `!pointee.is_freeze()`, but I'm afraid that this would cause the lint to miss more subtle cases where the type has interior mutability somewhere deeply nested.
@RalfJung what's your opinion on this here? Should the compiler emit a deny-by-default lint on this code below?

RalfJung replied:
I am totally fine with linting against this, calling `get` is a lot more clear and the docs for UnsafeCell explicitly say this is UB. It might be "just" library UB instead of language UB, but that's still good enough to justify a lint IMO. We can always revisit this once we have an official aliasing model.

Nilstrieb commented on compiler/rustc_lint/src/reference_casting.rs:
```diff
@@ -0,0 +1,120 @@
+fn is_ref_cast_from_const_to_mut<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'tcx>) -> bool {
+    let e = e.peel_blocks();
+    let e = if let ExprKind::Cast(e, t) = e.kind
+        && let TyKind::Ptr(MutTy { mutbl: Mutability::Mut, .. }) = t.kind {
```
I don't like how this matches on `hir::TyKind`, this can miss cases, like when people use `as _` (although in this particular case I can't get it to fail since inference doesn't seem to allow it). Can you instead get the `typeck_results.expr_ty(e)` and match on that? Same below.

Urgau replied:
Done, although I had to use [`node_type`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.node_type) directly since `t` is `hir::Ty` and not `hir::Expr` as expected by `expr_ty`.

Nilstrieb commented on compiler/rustc_lint/src/reference_casting.rs:
```diff
@@ -0,0 +1,120 @@
+    }
+}
+
+fn is_ref_cast_from_const_to_mut<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'tcx>) -> bool {
```
This method does something conceptually simple but the implementation is a little complicated due to the HIR matching. I'd like to see a few comments about what exactly is matched by each step (like you did in `check_expr`, that's lovely :3).

Urgau replied:
Done.

Noratrieb reviewed (changes requested):
I have some feedback and a question around `UnsafeCell`.

The original PR will land in 1.72.0 (nightly, beta in a few days). The rename should really get into the same release, so can you split it out into a separate PR to avoid it being caught up in bikeshedding here?

bors commented:
:umbrella: The latest upstream changes (presumably #111717) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Urgau committed:
Revert "Temporarily switch invalid_reference_casting lint to allow-by-default"

This reverts commit f25ad54a4d0febbcb2b7e951835228b7b2320b49.

Urgau committed:
Add support for deferred casting for the invalid_reference_casting lint

Urgau committed:
Avoid linting on expression that are only UB with SB/TB

Urgau committed:
Improve diagnostics of the invalid_reference_casting lint

Urgau commented:
Rebased on top of https://github.com/rust-lang/rust/pull/113422 and cleaned the commits.

@rustbot label +S-blocked
*(until https://github.com/rust-lang/rust/pull/113422 is merged)*

Urgau commented:
https://github.com/rust-lang/rust/pull/113422 has now been merged. I've rebased. This PR is no longer blocked and is ready for another review.

@rustbot label: -S-blocked +S-waiting-on-review

Nilstrieb commented:
@bors r+

bors commented:
:hourglass: Testing commit 507d497cfa014514c5dd904f9d273810e888f757 with merge d1708334312f654204edd21b3e93cf29da847187...
<!-- homu: {"type":"BuildStarted","head_sha":"507d497cfa014514c5dd904f9d273810e888f757","merge_sha":"d1708334312f654204edd21b3e93cf29da847187"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5738217156/job/15555441078)
Approved by: Nilstrieb
Pushing d1708334312f654204edd21b3e93cf29da847187 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Nilstrieb","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5738217156/job/15555441078"},"merge_sha":"d1708334312f654204edd21b3e93cf29da847187"} -->

rust-timer commented:
Finished benchmarking commit (d1708334312f654204edd21b3e93cf29da847187): [comparison URL](https://perf.rust-lang.org/compare.html?start=64ad036307727d21070a46f1a5e323cd442229da&end=d1708334312f654204edd21b3e93cf29da847187&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=64ad036307727d21070a46f1a5e323cd442229da&end=d1708334312f654204edd21b3e93cf29da847187&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.4%, 0.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=64ad036307727d21070a46f1a5e323cd442229da&end=d1708334312f654204edd21b3e93cf29da847187&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.2% | [-2.2%, -2.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-4.6%, -2.5%] | 2     |
| All ❌✅ (primary)                 | -2.2% | [-2.2%, -2.2%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=64ad036307727d21070a46f1a5e323cd442229da&end=d1708334312f654204edd21b3e93cf29da847187&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=64ad036307727d21070a46f1a5e323cd442229da&end=d1708334312f654204edd21b3e93cf29da847187&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 647.763s -> 648.737s (0.15%)
